src/CodeGen_WebGPU_Dev.cpp

```diff
@@ -199,6 +213,16 @@ string CodeGen_WebGPU_Dev::CodeGen_WGSL::print_type(Type type,
     return oss.str();
 }
 
+string CodeGen_WebGPU_Dev::CodeGen_WGSL::print_buffer_type(Type type) {
+    if (type.bits() != 32) {
+        internal_assert(type.bits() == 8 || type.bits() == 16);
+        internal_assert(!type.is_float());
+        // Types that are not 32-bits are emulated using atomics.
```

Comment:
What's the performance hit from using atomics in this way?

Reply:
Very substantial. Depends on the platform, the access pattern, and the amount of data, but we've measured anything from 2X to >10X slowdown compared to just using 32-bit types instead. As soon as WGSL gains an extension for these types we can start using that instead, which will help on platforms that can do this natively (macOS, Linux/Android with appropriate Vulkan extensions, and Windows with SM 6+).

If the slowdown is too scary to expose this by default, we could hide this functionality behind a Halide feature flag (e.g. `webgpu-int16`), which could use the extension unconditionally (and produces a runtime error if you try and run those pipelines on a platform that doesn't have it).

Also worth noting: we could avoid the atomics for cases where the accesses are dense loads/stores of whole vectors, meaning we would only incur this penalty for scalar i8/i16 stores or scatters/gathers. Not sure if that helps.

Reply:
> As soon as WGSL gains an extension for these types we can start using that instead, which will help on platforms that can do this natively (macOS, Linux/Android with appropriate Vulkan extensions, and Windows with SM 6+).

Any ETA for when that extension might exist?
 
> If the slowdown is too scary to expose this by default

...it looks pretty scary, I gotta admit
 
IIRC @abadams mentioned that other backends have ways to fake smaller types, but I don't know how they are implemented.

Reply:
I've been under the impression that the D3D backend does the same emulation via atomics:
https://github.com/halide/Halide/blob/master/src/CodeGen_D3D12Compute_Dev.cpp#L446-L479

But now that I look closer, it looks like this codepath is actually unused. It appears that the D3D backend can do this without atomics, by just setting the format at the buffer view to something like `DXGI_FORMAT_R16_UINT` on the runtime side whilst declaring the buffer as a 32-bit integer in the shader. This is all news to me, so let me think about whether this changes any of our thoughts about these types on the WebGPU/Chrome side.

Reply:
Adding @slomp and @shoaibkamil for the questions about the D3D backend

Reply:
Yup, in the case of D3D12, it's possible to define a "buffer view" with a specific format over the buffer memory, then "sample" it from the shader code and the hardware will pack/unpack the values accordingly to either `float` or `int`/`uint` (these are the only native types supported by HLSL, disregarding Shader Model 6.2).

Reply:
(In GL-land, you can achieve that with what ARB refers to as "texture buffer".)